MATLAB/rl_siso_NEW.m

```diff
@@ -0,0 +1,496 @@
+
+% Insert the cyclic prefix
+if(CP_LEN > 0)
+    % Do yourselves: Insert CP
+    tx_cp = tdd_tx_payload_mat((end-CP_LEN+1 : end), :);
+    tdd_tx_payload_mat = [tx_cp; tdd_tx_payload_mat];
+end
+
+% Reshape to a vector
+tx_payload_vec = reshape(tdd_tx_payload_mat, 1, numel(tdd_tx_payload_mat));
+
+
+% Construct the full time-domain OFDM waveform
+MAX_N_SAMPS = 4096;    % Maximum number of samples we can write to the FPGA buffer for TX
+N_ZPAD_POST = MAX_N_SAMPS - N_ZPAD_PRE - length(preamble) - length(tx_payload_vec);
+tx_vec = [zeros(1,N_ZPAD_PRE) preamble tx_payload_vec zeros(1,N_ZPAD_POST)];
+
+% Leftover from zero padding:
+tx_vec_iris = tx_vec.';
+% Scale the Tx vector to +/- 1
+tx_vec_iris = TX_SCALE .* tx_vec_iris ./ max(abs(tx_vec_iris));
+
+if SIM_MODE
+    disp("Running: AWGN SIMULATION MODE");
+    % AWGN only
+    snr = 20;
+    tx_var = mean(mean(abs(tx_vec_iris).^2 )) * (64/48);
+    nvar =  tx_var / 10^(0.1*snr); % noise variance per data sample
+        
+    H_ul = ones(size(tx_vec_iris.'));
+    % noise vector
+    W_ul = sqrt(nvar/2).* (randn(N_BS_NODE, length(tx_vec_iris)) + ...
+        1i*randn(N_BS_NODE, length(tx_vec_iris)) );
+    
+    % output vector
+    rx_vec_iris_tmp = H_ul.*tx_vec_iris.' + W_ul;
+    rx_vec_iris_tmp = rx_vec_iris_tmp.';
+
+    numGoodFrames = 1;  % Don't care about number of frames in SIM MODE
+else
+    % Set up the Iris experiment
+    disp("Running: HARDWARE MODE");
+
+    % Create two Iris node objects:
+    tx_direction = 'uplink';      % Options: {'uplink', 'downlink', 'hub_sync'}
+    bs_ids = ["RF3E000722"];
+    ue_ids = ["RF3E000665"];
+    hub_id = ["FH4B000003"];
+
+    % Iris nodes' parameters
+    sdr_params = struct(...
+        'bs_id', bs_ids, ...
+        'ue_id', ue_ids,...
+        'hub_id', hub_id,...
+        'bs_ant', ANT_BS, ...
+        'ue_ant', ANT_UE, ...
+        'txfreq', TX_FRQ, ...
+        'rxfreq', RX_FRQ, ...
+        'txgain', TX_GN, ...
+        'tx_gain_ue', TX_GN_UE, ...
+        'rxgain', RX_GN, ...
+        'sample_rate', SMPL_RT);
+
+    mimo_handle = mimo_driver(sdr_params);
+
+    if FIND_OPTIMAL_GAINS
+        [txg_opt, rxg_opt, max_num_beacon, valid] = mimo_handle.mimo_set_opt_gains(N_FRM);
+        fprintf(" !!! Optimal Gain Setting: TX %d, RX %d, max_num_beacon %d !!! \n", txg_opt, rxg_opt, max_num_beacon);
+        if ~valid
+            printf("[ERROR] Could NOT find an adequate gain setting (no beacons detected). Check setup!");
+            return;
+        end
+        mimo_handle.mimo_update_sdr_param('txgain', txg_opt);
+        mimo_handle.mimo_update_sdr_param('rxgain', rxg_opt);
+    end
+
+    if strcmp(tx_direction, 'uplink')
+        [rx_vec_iris_tmp, numGoodFrames, ~] = mimo_handle.mimo_txrx_uplink(tx_vec_iris, N_FRM, N_ZPAD_PRE);
+
+    elseif strcmp(tx_direction, 'downlink')
+        [rx_vec_iris_tmp, numGoodFrames, ~] = mimo_handle.mimo_txrx_downlink(tx_vec_iris, N_FRM, N_ZPAD_PRE);
+
+    elseif strcmp(tx_direction, 'hub_sync')
```

Comment:
`hub_sync` name is confusing. Perhaps something to indicate direction and the sync mode?

Reply:
changed it